Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,613 @@
+
+        /// Creates a new `HTTP1ConnectionProvider`
+        ///
+        /// - parameters:
+        ///     - key: The `Key` (host, scheme, port) this provider is associated to
+        ///     - configuration: The client configuration used globally by all requests
+        ///     - initialConnection: The initial connection the pool initializes this provider with
+        ///     - parentPool: The pool this provider belongs to
+        init(key: ConnectionPool.Key, configuration: HTTPClient.Configuration, initialConnection: Connection, parentPool: ConnectionPool) {
+            self.eventLoop = initialConnection.channel.eventLoop
+            self.configuration = configuration
+            self.key = key
+            self.parentPool = parentPool
+            self.isClosed = NIOAtomic.makeAtomic(value: false)
+            self.state = State(initialConnection: initialConnection)
+            self.configureCloseCallback(of: initialConnection)
+        }
+
+        deinit {
+            // FIXME: Add the checks back
+            /*
+             assert(availableConnections.isEmpty, "Available connections should be empty before deinit")
+             assert(leased == 0, "All leased connections should have been returned before deinit")
+             */
+        }
+
+        func getConnection(preference: HTTPClient.EventLoopPreference) -> EventLoopFuture<Connection> {
+            let action = self.stateLock.withLock { self.state.connectionAction(for: preference) }
+            switch action {
+            case .leaseConnection(let connection):
+                return self.eventLoop.makeSucceededFuture(connection)
+            case .makeConnection(let eventLoop):
+                return self.makeConnection(on: eventLoop)
+            case .leaseFutureConnection(let futureConnection):
+                return futureConnection
+            }
+        }
+
+        func release(connection: Connection) {
+            let action = self.stateLock.withLock { self.state.releaseAction(for: connection) }
+            switch action {
+            case .succeed(let promise):
+                promise.succeed(connection)
+
+            case .makeConnectionAndComplete(let eventLoop, let promise):
+                self.makeConnection(on: eventLoop).cascade(to: promise)
+
+            case .replaceConnection(let eventLoop):
+                connection.channel.close().flatMap {
+                    self.makeConnection(on: eventLoop)
+                }.whenComplete { result in
+                    switch result {
+                    case .success(let connection):
+                        self.release(connection: connection)
+                    case .failure(let error):
+                        let (promise, providerMustClose) = self.stateLock.withLock { self.state.popConnectionPromiseToFail() }
+                        promise?.fail(error)
+                        if providerMustClose {
+                            self.parentPool[self.key] = nil
```

Comment:
I think this is some of the synchronisation stuff @weissi was worried about with the subscript on the connection pool. While destroying a provider is fine, there is no guarantee that the provider we're about to destroy is _this one_. Maybe we already destroyed ourselves in another thread and since created a new provider for this key (note that multiple threads may both find that they want to close this provider).

In general it may be better for all subscript accessors to the pool to instead be rewritten in terms of methods that explicitly say what their semantic should be (i.e. "remove me from the pool if I'm in it") and that manage the locks appropriately to ensure that this operation works as intended.

Reply:
With the latest changes where we check that a connection pool isn't reused, I'm indeed able to see there is issues linked to this, will fix this, thanks!